src/espLoader.ts

```diff
@@ -0,0 +1,1008 @@
+import { ESPError } from "./error";
+import { inflate, deflate } from "pako";
+import ESP32ROM from "./targets/esp32";
+import { Transport } from "./webserial";
+import { ROM } from "./targets/rom";
+import ESP32C3ROM from "./targets/esp32c3";
+import ESP32S3ROM from "./targets/esp32s3";
+import ESP32S2ROM from "./targets/esp32s2";
+import ESP8266ROM from "./targets/esp8266";
+
+const MAGIC_TO_CHIP: { [key: number]: ROM } = {
+  0x00f01d83: new ESP32ROM(),
+  0x6921506f: new ESP32C3ROM(), // ECO 1+2
+  0x1b31506f: new ESP32C3ROM(), //ECO 3
+  0x09: new ESP32S3ROM(),
+  0x000007c6: new ESP32S2ROM(),
+  0xfff0c101: new ESP8266ROM(),
+};
+
+interface ESPBinFile {
+  data: string;
+  address: number;
+}
+
+class ESPLoader {
+  ESP_RAM_BLOCK = 0x1800;
+  ESP_FLASH_BEGIN = 0x02;
+  ESP_FLASH_DATA = 0x03;
+  ESP_FLASH_END = 0x04;
+  ESP_MEM_BEGIN = 0x05;
+  ESP_MEM_END = 0x06;
+  ESP_MEM_DATA = 0x07;
+  ESP_WRITE_REG = 0x09;
+  ESP_READ_REG = 0x0a;
+
+  ESP_SPI_ATTACH = 0x0d;
+  ESP_CHANGE_BAUDRATE = 0x0f;
+  ESP_FLASH_DEFL_BEGIN = 0x10;
+  ESP_FLASH_DEFL_DATA = 0x11;
+  ESP_FLASH_DEFL_END = 0x12;
+  ESP_SPI_FLASH_MD5 = 0x13;
+
+  // Only Stub supported commands
+  ESP_ERASE_FLASH = 0xd0;
+  ESP_ERASE_REGION = 0xd1;
+  ESP_RUN_USER_CODE = 0xd3;
+
+  ESP_IMAGE_MAGIC = 0xe9;
+  ESP_CHECKSUM_MAGIC = 0xef;
+
+  // Response code(s) sent by ROM
+  ROM_INVALID_RECV_MSG = 0x05; // response if an invalid message is received
+
+  ERASE_REGION_TIMEOUT_PER_MB = 30000;
+  ERASE_WRITE_TIMEOUT_PER_MB = 40000;
+  MD5_TIMEOUT_PER_MB = 8000;
+  CHIP_ERASE_TIMEOUT = 120000;
+  MAX_TIMEOUT = this.CHIP_ERASE_TIMEOUT * 2;
+
+  CHIP_DETECT_MAGIC_REG_ADDR = 0x40001000;
+
+  DETECTED_FLASH_SIZES: { [key: number]: string } = {
+    0x12: "256KB",
+    0x13: "512KB",
+    0x14: "1MB",
+    0x15: "2MB",
+    0x16: "4MB",
+    0x17: "8MB",
+    0x18: "16MB",
+  };
+
+  chip: ROM;
+  IS_STUB: boolean;
+  FLASH_WRITE_SIZE: number;
+
+  constructor(
+    public transport: Transport,
+    private baudrate: number,
+    private terminalCleanFn: () => void,
+    private terminalWriteLineFn: (data: string) => void,
+    private terminalWriteFn: (data: string) => void,
```

Comment:
I think that the old way that allowed passing in a terminal object was nicer. Easier to reason about. Why did you break it up?

Reply:
Because we want to use other terminals beside xTerm. For example, use this Esptool-js in Eclipse Theia and vscode extension. The previous implementation was using terminal functions within espLoader, I think injecting dependency gives more flexibility. Passing the entire Terminal object is also no go, because methods like `writeLn` are not the same for all terminal frameworks.

Reply:
Right, but it could be an interface that consumers can wrap their terminal in.

```ts
new ESPLoader({
  clean() {
    myTerm.doClean();
  }

  write(data) {
    myTerm.doWrite(data)
  }
})
```